EIPS/eip-9999.md

```diff
@@ -0,0 +1,150 @@
+2. All subsequent accesses to the same slot within the same block incur only the warm access cost as of [EIP-2929](./eip-2929).
+3. The warm/cold status resets at block boundaries
+
+### Block Processing
+
+1. At the start of each block:
+   * Initialize two empty sets `block_level_accessed_addresses` and `block_level_accessed_storage_keys`
+2. For each transaction in the block:
+   * Before processing storage access:
+     * Check if touched address is in `block_level_accessed_addresses`
+     * If yes: charge `GAS_WARM_ACCESS`
+     * If no:
+       * Charge `GAS_COLD_ACCOUNT_ACCESS`
+       * Add address to `block_level_accessed_addresses`
+     * Check if storage key is in `block_level_accessed_storage_keys`
+     * If yes: charge `GAS_WARM_ACCESS`
+     * If no:
+       * Charge `GAS_COLD_SLOAD`
+       * Add storage key to `block_level_accessed_storage_keys`
+    
+
+### Implementation Details
+
+The implementation modifies the block execution process to maintain block-level sets of accessed addresses and storage slots. 
+
+#### Block-Level Storage Management
+
+```python
+def apply_body(...):
+    # Initialize block-level tracking sets
+    block_level_accessed_addresses = set()
+    block_level_accessed_storage_keys = set()
+    
+    for i, tx in enumerate(map(decode_transaction, transactions)):
+        # Create environment with block-level context
+        env = vm.Environment(
+            # ... other parameters ...
+            block_level_accessed_addresses=block_level_accessed_addresses,
+            block_level_accessed_storage_keys=block_level_accessed_storage_keys
+        )
+        
+        # Process transaction and update block-level sets
+        gas_used, accessed_addresses, accessed_storage_keys, logs, error = process_transaction(env, tx)
+        block_level_accessed_addresses += accessed_addresses
+        block_level_accessed_storage_keys += accessed_storage_keys
+```
+
+This code shows how block-level slot warming is implemented at the execution layer. The `block_level_accessed_addresses` and `block_level_accessed_storage_keys` sets are maintained throughout the block's execution and passed to each transaction's environment.
+
+#### Transaction Processing
+
+```python
+def process_transaction(env: vm.Environment, tx: Transaction) -> Tuple[Uint, Tuple[Log, ...], Optional[Exception]]:
+    preaccessed_addresses = set()
+    preaccessed_storage_keys = set()
+    
+    # Add block-level pre-accessed slots
+    preaccessed_addresses.add(env.block_level_accessed_addresses)
+    preaccessed_storage_keys.add(env.block_level_accessed_storage_keys)
+    
+    # Handle access lists from transaction
+    ...
+```
+
+This adds the block-level-accessed addresses and storage keys to the preaccessed addresses and storage keys.
+As a result, from the perspective of a transaction, block-level accessed addresses and storage keys are treated the same as precompiles or the coinbase address.
```

Comment:
What if the transaction itself reverts? In EIP-2929 we have the logic that if a sub-call reverts, all the accessed addresses/slots are now cold. In case of a transaction failure, what happens here? Are all the accessed addresses/slots in this specific addresses also removed? Will put a point at the conclusion of this review if we should consider removing this revert logic altogether since that would fit within the scope/idea of this PR.

Reply:
This is a very good point. I agree that it might make sense to keep slots warm even if the tx reverts. Will add it to the rational for now